### PR TITLE
Multiple seat support

### PR DESCRIPTION
**Pull request by beausmith**
This isn't finished… because tests aren't updated… but it works.

### CODE REVIEW

**Comment by beausmith** [comment]
Here is a animated gif of the functionality…

multi-seat

**Comment by benadida** [comment]
very impressed with how well this works with write-ins too.

File: src/components/SeatContest.tsx [hunk 1]
```diff
@@ -236,17 +248,17 @@ class SeatContest extends React.Component<Props, State> {
...
-                {!!vote ? `1` : `0`}.
+                <strong>Vote for {contest.seats}.</strong> You have selected{' '}
+                {vote.length}.
               </p>
             </Prose>
           </Legend>
           <Choices>
             {contest.candidates.map((candidate, index) => {
-              const isChecked = !!vote && candidate.name === vote.name
+              const isChecked = !!vote.find(c => c.id === candidate.id)
```
Comment (benadida) [inline src/components/SeatContest.tsx]: here's that `find` again :)

File: src/App.tsx [hunk 2]
```diff
@@ -113,9 +113,9 @@ class App extends React.Component<RouteComponentProps, State> {
     this.props.history.push('/')
   }
 
-  public updateVote = (contestId: string, candidate: OptionalCandidate) => {
+  public updateVote = (contestId: string, vote: OptionalVote) => {
```
[the remark this answers is not in the document]
Comment (beausmith) [reply src/App.tsx]: If confusing, let's update the terminology. Please propose any terms you find less confusing.

Here's my current thinking…

It hasn't been defined how a voter's "ballot" will be distilled from an election, but I've been thinking that `this.state.ballot` would be used for this. 

Possible `App` state:

```
{
  election: Election // <-- election config json
  ballot: Ballot // <-- TBD, created when an activation code is entered.
  votes: Votes {
    contestId: Candidate[] || Approval // <-- Approval will be a boolean type
  }
}
```

If the voter's "ballot" is dynamically derived from the election (instead of creating a ballot when an activation code is entered) then we can change `votes` to `ballot`.

File: src/components/SeatContest.tsx [hunk 3]
```diff
@@ -125,101 +128,110 @@ class SeatContest extends React.Component<Props, State> {
...
 
-  public selectCandidate = (candidate: OptionalCandidate) => {
-    this.props.updateVote(this.props.contest.id, candidate)
+  public addCandidateToVote = (id: string) => {
+    const { contest, vote = [] } = this.props
+    const { candidates } = contest
+    const candidate = candidates.find(c => c.id === id)
+    if (candidate) {
+      this.props.updateVote(contest.id, [...vote, candidate])
+    }
```
[the remark this answers is not in the document]
Comment (beausmith) [reply src/components/SeatContest.tsx]: Hehe! Code coverage report (which I haven't completed yet) will complain that the else path is not followed. Ok I'll look into this.

File: src/components/SeatContest.tsx [hunk 4]
```diff
@@ -125,101 +128,110 @@ class SeatContest extends React.Component<Props, State> {
...
-    const { value } = event.target as HTMLInputElement
-    const targetIsSelected = vote && value === vote.id
-    this.selectCandidate(
-      targetIsSelected
-        ? undefined
-        : contest.candidates.find(candidate => candidate.id === value)
-    )
+    const { vote = [] } = this.props
+    const id = (event.target as HTMLInputElement).value
+    const candidate = vote.find(c => c.id === id)
```
[the remark this answers is not in the document]
Comment (beausmith) [reply src/components/SeatContest.tsx]: Done.

File: src/components/SeatContest.tsx [hunk 5]
```diff
@@ -125,101 +128,110 @@ class SeatContest extends React.Component<Props, State> {
...
+    if (!!candidate) {
+      if (candidate.writeIn) {
+        this.setState({ candidatePendingRemoval: candidate })
```
[the remark this answers is not in the document]
Comment (beausmith) [reply src/components/SeatContest.tsx]: If the voter took the time to add a write-in, and then accidentally clicks to uncheck and the write-in is gone without any warning confirmation… I think they would be disappointed. Thinking about older people fat-fingering the touchscreen. 😄

File: src/components/SeatContest.tsx [hunk 6]
```diff
@@ -125,101 +128,110 @@ class SeatContest extends React.Component<Props, State> {
...
-  }
-
   public onKeyboardInputChange = (writeInCandidateName: string) => {
     this.setState({ writeInCandidateName })
   }
 
   public render() {
-    const { contest, vote } = this.props
+    const { contest, vote = [] } = this.props
+    const seatsAreFull = vote.length >= contest.seats
```
[the remark this answers is not in the document]
Comment (beausmith) [reply src/components/SeatContest.tsx]: Abstracted.

File: src/components/SeatContest.tsx (hunk 1 above)
Comment (beausmith) [reply src/components/SeatContest.tsx]: Abstracted.

File: src/components/SeatContest.tsx [hunk 7]
```diff
@@ -125,101 +128,110 @@ class SeatContest extends React.Component<Props, State> {
...
+    if (candidate) {
+      this.props.updateVote(contest.id, [...vote, candidate])
+    }
+  }
+
+  public removeCandidateFromVote = (id: string) => {
+    const { contest, vote = [] } = this.props
+    if (id) {
+      this.props.updateVote(contest.id, vote.filter(c => c.id !== id))
+    }
```
[the remark this answers is not in the document]
Comment (beausmith) [reply src/components/SeatContest.tsx]: Done and abstracted.

**Comment by beausmith** [comment]
@benadida I believe I have addressed all of your comments. I'm clicking the link to request a re-review.

**Comment by beausmith** [comment]
Looking into failed tests…

**Comment by beausmith** [comment]
Fixed failing tests.

**Comment by beausmith** [comment]
Now the tests are really fixed! 🤞

File: public/data/election.json [hunk 8]
```diff
@@ -6,25 +6,37 @@
   "seal": "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 500 500\">\n  <circle cx=\"250\" cy=\"250\" r=\"240\" fill=\"#fff\" stroke=\"#000\" stroke-width=\"10\"/>\n  <text x=\"50%\" y=\"50%\" text-anchor=\"middle\" alignment-baseline=\"middle\" style=\"font-family: sans-serif; font-size: 180%\">Official Seal Here</text>\n</svg>\n",
   "contests": [
     {
-      "id": "president",
+      "allowWriteIn": true,
```
Comment (benadida) [inline public/data/election.json]: as per discussion, `allowWriteIns`.

File: src/components/SeatContest.tsx [hunk 9]
```diff
@@ -125,101 +132,110 @@ class SeatContest extends React.Component<Props, State> {
...
     )
-    this.setState({ writeInCandidateName })
-    this.selectCandidate({
-      id: 'writeInCandidate',
-      name: writeInCandidateName,
-    })
+    this.props.updateVote(contest.id, [
+      ...vote,
+      {
+        id: camelCase(normalizedCandidateName),
```
Comment (benadida) [inline src/components/SeatContest.tsx]: I think the `id` here should probably be `write-in-1`, `write-in-2`, etc. I can imagine collisions between write-in names and existing candidates, and in general I think it may be a potential problem to use IDs that are decided by user input.

File: src/components/SeatContest.tsx [hunk 10]
```diff
@@ -275,39 +292,58 @@ class SeatContest extends React.Component<Props, State> {
...
-          isOpen={!!attemptedVoteCandidateName}
+          isOpen={!!attemptedOverVoteCandidate}
           content={
             <Prose>
-              <Text>
-                To vote for {attemptedVoteCandidateName}, first uncheck the vote
-                for {!!vote && vote.name}.
-              </Text>
+              {contest.seats === 1 ? (
+                <Text>
```
Comment (benadida) [inline src/components/SeatContest.tsx]: nit: don't repeat what you don't need to repeat, e.g. `<Text>`.

Comment (beausmith) [reply src/components/SeatContest.tsx]: Simplified.

File: src/components/SeatContest.tsx (hunk 9 above)
Comment (beausmith) [reply src/components/SeatContest.tsx]: Good call. Fixed by prefixing camelCased candidate name with `write-in__`.

**Comment by beausmith** [comment]
@benadida Ready for re-re-re-review.

File: src/App.test.tsx [hunk 11]
```diff
@@ -133,107 +129,147 @@ it('end to end: election can be uploaded, voter can vote and print', async () =>
...
-  fireEvent.click(getByText(contest0Candidate0).closest('label')!)
-  fireEvent.click(getByText('add a write-in candidate').closest('label')!)
-  fireEvent.click(getByText('Close'))
-  fireEvent.click(getByText('add a write-in candidate').closest('label')!)
+  expect(getByText(multiSeatCandidate0.name)).toBeTruthy()
+  expect(getByText(multiSeatCandidate1.name)).toBeTruthy()
+  expect(getByText(multiSeatCandidate2.name)).toBeTruthy()
+  expect(getByText(multiSeatCandidate3.name)).toBeTruthy()
+
+  // Test Write-In Candidate flow
```
Comment (benadida) [inline src/App.test.tsx]: awesome, full commitment to all test paths :)

File: src/components/Ballot.test.tsx [hunk 12]
```diff
@@ -49,7 +49,7 @@ it('redirects contests index to first contest', () => {
   const { container, getByText } = render(<Ballot />, {
     route: '/contests',
   })
-  getByText('President and Vice-President of the United States')
+  getByText(electionSample.contests[0].title)
```
Comment (benadida) [inline src/components/Ballot.test.tsx]: 👍

File: src/components/CandidateContest.tsx [hunk 13]
```diff
@@ -125,101 +132,110 @@ class CandidateContest extends React.Component<Props, State> {
...
-    })
+    this.props.updateVote(contest.id, [
+      ...vote,
+      {
+        id: `write-in__${camelCase(normalizedCandidateName)}`,
+        isWriteIn: true,
+        name: normalizedCandidateName,
+      },
+    ])
+    this.setState({ writeInCandidateName: '' })
```
Comment (benadida) [inline src/components/CandidateContest.tsx]: I'm still struggling with this empty string as null situation, but I'll leave it be until I have a better suggestion.

File: src/data/electionSample.json [hunk 14]
```diff
@@ -554,37 +554,37 @@
...
...mber of any board, commission or other state or quasi-public entity exercising executive power under the laws of this state shall be a member of the senate or the house of representatives during his or her continuance in such office.\n\nb. Article V of the Constitution shall be amended to read as follows: The powers of the government shall be distributed into three (3) separate and distinct departments: the legislative, the executive and the judicial.\n\nc. Article VI, Section 10 of the Constitution shall be deleted in its entirety.\n\nd. Article IX, Section 5 of the Constitution shall be amended to read as follows:\n\nSection 5. Powers of appointment.- The governor shall, by and with the advice and consent of the senate, appoint all officers of the state whose appointment is not herein otherwise provided for and all members of any board, commission or other state or quasi-public entity which exercises executive power under the laws of this state; but the general assembly may by law vest the appointment of such inferior officers, as they deem proper, in the governor, or within their respective departments in the other general officers, the judiciary or in the heads of departments."
```
Comment (benadida) [inline src/data/electionSample.json]: I think we decided not  to test the super long proposition right now. Are you sure you want to force this now? Feels like an edge case that could be tricky.

Comment (beausmith) [reply src/data/electionSample.json]: Oh, I'm not aware of that decision. Please tell me more. 

My reasoning: It was in the sample content, so I assumed others were expecting it. Also long content is unavoidable unless we enforce it (when creating or loading json config). Since one of the next tasks is to add scroll buttons, I need long content anyhow. 

Happy to adjust.
